Mathlib/Combinatorics/Nullstellensatz.lean

```diff
@@ -0,0 +1,315 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+import Mathlib.Algebra.MvPolynomial.Equiv
+import Mathlib.Algebra.Polynomial.Degree.Definitions
+import Mathlib.Data.Finsupp.MonomialOrder.DegLex
+import Mathlib.Data.Set.Card
+import Mathlib.RingTheory.Ideal.Maps
+import Mathlib.RingTheory.MvPolynomial.Groebner
+import Mathlib.RingTheory.MvPolynomial.Homogeneous
+import Mathlib.RingTheory.MvPolynomial.MonomialOrder.DegLex
+
+/-! # Alon's Combinatorial Nullstellensatz
+
+This is a formalization of Noga Alon's Combinatorial Nullstellensatz. It follows [Alon_1999].
+
+We consider a family `S : σ → Finset R` of finite subsets of a domain `R`
+and a multivariate polynomial `f` in `MvPolynomial σ R`.
+The combinatorial Nullstellensatz gives combinatorial constraints for
+the vanishing of `f` at any `x : σ → R` such that `x s ∈ S s` for all `s`.
+
+- `MvPolynomial.eq_zero_of_eval_zero_at_prod` :
+  if `f` vanishes at any such point and `f.degreeOf s < (S s).ncard` for all `s`,
+  then `f = 0`.
+
+- `Polynomial.eq_zero_of_eval_zero` :
+  a polynomial (in one indeterminate) which vanishes at more points than its degree
+  is the zero polynomial.
+
+- `combinatorial_nullstellensatz_exists_linearCombination`
+  If `f` vanishes at every such point, then it can be written as a linear combination
+  `f = linearCombination (MvPolynomial σ R) (fun i ↦ (∏ r ∈ S i, (X i - C r))) h`,
+  for some `h : σ →₀ MvPolynomial σ R` such that
+  `((∏ r ∈ S s, (X i - C r)) * h i).totalDegree ≤ f.totalDegree` for all `s`.
+
+- `combinatorial_nullstellensatz_exists_eval_nonzero`
+  a multi-index `t : σ →₀ ℕ` such that `t s < (S s).card` for all `s`,
+  `f.totalDegree = t.degree` and `f.coeff t ≠ 0`,
+  there exists a point `x : σ → R` such that `x s ∈ S s` for all `s` and `f.eval s ≠ 0`.
+
+## TODO
+
+- Applications
+- relation with Schwartz–Zippel lemma, as in [Rote_2023]
+
+## References
+
+- [Alon, *Combinatorial Nullstellensatz*][Alon_1999]
+
+- [Rote, *The Generalized Combinatorial Lason-Alon-Zippel-Schwartz
+  Nullstellensatz Lemma*][Rote_2023]
+
+-/
+
+open Finsupp
+
+variable {R : Type*} [CommRing R]
+
+/-- A polynomial that vanishes at more points than its degree is the zero polynomial. -/
+theorem _root_.Polynomial.eq_zero_of_eval_zero [IsDomain R] (P : Polynomial R) (S : Set R)
+    (Hdeg : Polynomial.natDegree P < S.ncard) (Heval : ∀ x ∈ S, P.eval x = 0) :
+    P = 0 := by
```

Comment:
I feel like this lemma belongs in an earlier file. Can you try moving?

Reply:
There's an issue with `Set.ncard` that isn't accessible in files such as `Polynomial.Roots`.
On the other hand, I could as well work with `S : Finset R`, and then the issue would disappear.

Reply:
I agree in principle that this lemma and the one in several variables (for product sets) belong elsewhere.

Reply:
> There's an issue with `Set.ncard` that isn't accessible in files such as `Polynomial.Roots`.

What happens if you import it there? Does the number of transitive imports explode? (I would hope not, but I don't know)

Reply:
I'll do that. I don't know how to get the count of the transitive imports.

Reply:
Contreordre. I didn't do that. I observed that the finset case already existed and moved my lemmas to finset.
In fact, this is the context of the Comb.NS.

Reply:
> I'll do that. I don't know how to get the count of the transitive imports.

I'd say push and wait for the bot to yell at you on the PR